### PR TITLE
Support for two or more authentications for a single request

### LINKED ISSUE #31328: Support for inclusive authentication
### Description

If more than one authentication mechanism is registered then the authentication is complete as soon as the first `SecurityIdentity` is created by one of the mechanisms. However it is not enough when tokens or other credentials have to be carried over `mTLS` since only `mTLS` mechanism will create `SecurityIdentity`.

Quarkus Security should support `inclusive` authentication which will require all the registered authentication mechanisms to perform the authentication.

This enhancement is a prerequisite for #4482 (OIDC bearer access token binding to the client certificate) but I believe it will also be of general interest to support the user authentication over VPN, etc.

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by michalvavrik**
- Closes: https://github.com/quarkusio/quarkus/issues/31328
- Supersedes: https://github.com/quarkusio/quarkus/pull/31329
- many parts (OIDC, docs, tests) are just taken from Sergey's PR and tweaked
- main difference to https://github.com/quarkusio/quarkus/pull/31329 is that I addressed Stuarts concerns with new approach in Vert.X HTTP

### CODE REVIEW

**Comment by michalvavrik** [comment]
> io.quarkus.it.oidc.OidcMtlsTest.testGetIdentityNames

Interesting, I have run that test like 10 times, including native mode, never failed. I am going to look what is different in CI.

**Comment by michalvavrik** [comment]
So what has changed is that I am getting exception here https://github.com/quarkusio/quarkus/blob/main/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcProviderClient.java#L176 for OIDC client is is now missing.

**Comment by michalvavrik** [comment]
@sberyozkin do you think you could investigate failures? they are caused by https://github.com/quarkusio/quarkus/pull/41866 and I don't see anything wrong about OIDC setup in this PR. You have way better chance to figure it quickly as I think it might be a bug introduced in https://github.com/quarkusio/quarkus/pull/41866. Thank you

**Comment by sberyozkin** [comment]
Hi @michalvavrik I'm pretty sure the OIDC client reg work is unrelated - it is only about registering client dynamically, in fact, independently of `quarkus-oidc`. My bet is there is a race condition somewhere :-)

**Comment by sberyozkin** [comment]
@michalvavrik Major thanks by the way for picking up the old PR and concluding this work, it will be a very important feature

**Comment by michalvavrik** [comment]
> Hi @michalvavrik I'm pretty sure the OIDC client reg work is unrelated - it is only about registering client dynamically, in fact, independently of `quarkus-oidc`. My bet is there is a race condition somewhere :-)

Hmm, but I reverted that registration and this worked like a charm. I really think that your PR is related, but if you want, I can investigate it myself. NP, I just thought you are better poised to it. The fact is that this line not throws exception and it didn't: https://github.com/quarkusio/quarkus/blob/main/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcProviderClient.java#L176

I can investigate though.

> @michalvavrik Major thanks by the way for picking up the old PR and concluding this work, it will be a very important feature

I actually enjoyed it, it was nice little work that had to be done in Vert.X HTTP, nothing else :-)

File: docs/src/main/asciidoc/security-authentication-mechanisms.adoc [hunk 1]
```diff
@@ -567,6 +567,68 @@ ifndef::no-quarkus-elytron-security-oauth2[ ^|Yes]
...
+
+In Quarkus such authentication is called `inclusive` and you can enable it like this:
+
+[source,properties]
+----
+quarkus.http.auth.inclusive=true
+----
+
+If the authentication is inclusive then `SecurityIdentity` created by the first authentication mechanism can be
+injected into the application code. Typically it will be created by the <<mutual-tls>> authentication mechanism.
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-authentication-mechanisms.adoc:586]:
```suggestion
injected into the application code. For example, if both <<mutual-tls>> and basic authentication mechanisms authentications are required, the <<mutual-tls>> authentication mechanism will create `SecurityIdentity` first.
```

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/MtlsAuthenticationMechanism.java [hunk 2]
```diff
@@ -83,6 +83,11 @@ public Uni<HttpCredentialTransport> getCredentialTransport(RoutingContext contex
         return Uni.createFrom().item(new HttpCredentialTransport(HttpCredentialTransport.Type.X509, "X509"));
     }
 
+    @Override
```
Comment (sberyozkin) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/MtlsAuthenticationMechanism.java:86]: I think if we have inclusive authentication, then `HttpAuthenticator` must throw ISE if, after sorting enabled mechanisms, MTLS is in the list of the mechanisms but is not the first one. It has to be predictable IMHO, otherwise for ex, #4482 won't work. MTLS is the closest to the transport, so it won't be wrong to insist on it coming up first in case of the inclusive authentication. I'm on;y proposing to have this extra check, even though `3000` should guarantee it being first in most cases

Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/MtlsAuthenticationMechanism.java:86]: Okay, I think it sounds reasonable considering, as you said, is actually enabled and present. I'll add it.

**Comment by sberyozkin** [comment]
@michalvavrik May  be https://github.com/quarkusio/quarkus/pull/41866/files#diff-67404a43520250b1ebd5a0c945e070e60d70d4a3ba06b15e3fac747f820320cc are related though I'm not seeing how, I had to tweak the conditions for starting Keycloak Dev Services (to support starting even if the default tenant is disabled, and avoid registering default client). It looks like the default client is not registered by dev service

**Comment by sberyozkin** [comment]
@michalvavrik 

It must be this one:

https://github.com/quarkusio/quarkus/pull/41866/files#diff-67404a43520250b1ebd5a0c945e070e60d70d4a3ba06b15e3fac747f820320ccR855

Though again, I'm not sure exactly now why I did it

**Comment by michalvavrik** [comment]
> @michalvavrik
> 
> It must be this one:
> 
> https://github.com/quarkusio/quarkus/pull/41866/files#diff-67404a43520250b1ebd5a0c945e070e60d70d4a3ba06b15e3fac747f820320ccR855
> 
> Though again, I'm not sure exactly now why I did it

Cool, I knew I'll figure it out :-) Thanks for having a look after all.

**Comment by sberyozkin** [comment]
@michalvavrik Let me try to run OIDC client reg tests without those client id/secret updates to remind myself why I did it

**Comment by sberyozkin** [comment]
#42947 has been merged, please rebase

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/MtlsAuthenticationMechanism.java (hunk 2 above)
Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/MtlsAuthenticationMechanism.java:87]: Added validation and test.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java [hunk 3]
```diff
@@ -7,17 +7,44 @@
...
+    }
+
+    /**
+     * When inclusive authentication is enabled, we allow all authentication mechanisms to produce identity.
+     * However, only the first identity (provided by applicable mechanism with the highest priority) is stored
+     * in the CDI container. Therefore, we put all the identities into the RoutingContext.
+     *
+     * @return null if no identities were found or map with authentication mechanism key and security identity value
+     */
+    public static Map<String, SecurityIdentity> getSecurityIdentities(RoutingContext routingContext) {
```
[the remark this answers is not in the document]
Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java:44]: did you read `@return` javadoc? literally explains it :-) now, I am not being impertinent, I'd just like to understand if the text is not understandable and I should rephrase it, or if you mean to explain in more detail etc.

**Review by sberyozkin — APPROVED** [review]
Looks perfect to me. For OIDC Mutual TLS binding work, I can get a map of current security identities, get the MTLS one and extract the required data from the certificate to match them against the token

**Comment by sberyozkin** [comment]
@michalvavrik Thanks, only minor JavaDocs comment, indeed, lets see if Stuart can comment during the next couple of days

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java:44]: @michalvavrik Ouch, I did but only the first couple of words :-)

[the remark this answers is not in the document]
Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java:44]: yeah, it's not the best sentence, but it's there :-)

**Comment by sberyozkin** [comment]
Let's merge now

File: extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/security/InclusiveAuthValidationTest.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                    .addAsResource(new File("target/certs/mtls-test-server-truststore.jks"), "server-truststore.jks"))
+            .assertException(throwable -> {
+                var errMsg = throwable.getMessage();
+                Assertions.assertTrue(errMsg.contains("Inclusive authentication is enabled"));
+                Assertions.assertTrue(errMsg.contains("TopPriorityAuthMechanism"));
+            });
+
+    @Test
+    public void test() {
+        Assertions.fail();
```
Comment (rsvoboda) [inline extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/security/InclusiveAuthValidationTest.java:63]: @michalvavrik why is this here?

Comment (rsvoboda) [reply extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/security/InclusiveAuthValidationTest.java:63]: Sure, I noticed `assertException` on line 55.

It would be better to have a fail message for faster evaluation in case of fail.

Comment (michalvavrik) [reply extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/security/InclusiveAuthValidationTest.java:63]: ok, I'll think of it next time; I just expect that folk who debug / read it knows it, my bad

Comment (rsvoboda) [reply extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/security/InclusiveAuthValidationTest.java:63]: Sure, this is just minor thingy
